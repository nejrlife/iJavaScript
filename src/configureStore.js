import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';
import app from './reducers';

import dataSaga from './sagas';

const sagaMiddleware = createSagaMiddleware({
  onError: (error, action) => {
    if (error) {
      console.log('ERROR');
      console.log(error);
      console.log(action);
    }
  }
});

export default function configureStore() {
  const store = createStore(app, applyMiddleware(sagaMiddleware));
  sagaMiddleware.run(dataSaga);
  return store;
}