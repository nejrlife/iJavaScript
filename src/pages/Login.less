.loginBody {
  width: 100%;
  height: 100vh;
}
.loginContainer {
  width: 1008px;
  height: 330px;
  background-color: #CC0000;
  padding-top: 20px;
}
.errorMessage {
  width: 90%;
  margin-left: auto;
  margin-right: auto;
  background-color: #DD5656;
}
.errorMessage p {
  margin: 0;
  padding: 20px;
  color: #FFFFFF;
  font-style: italic;
}
.errorMessage p b {
  font-weight: 600;
}
.formContainer {
  margin-top: 20px;
  display: flex;
  flex-direction: column;
  margin-left: auto;
  margin-right: auto;
  width: 90%;
}
.formContainer label{
  width: 100px;
  font-family: Microsoft JhengHei;
  font-size: 18px;
  color: #FFF;
}
.formContainer label p {
  margin: 0px;
}
.formContainer input{
  margin-left: 20px;
  height: 40px;
  padding: 5px;
}
.formContainer ::placeholder {
  color: gray;
  opacity: 0.5; /* Firefox */
  font-family: Microsoft JhengHei;
  font-size: 18px;
}
.formContainer input::-webkit-autofill::first-line {
  font-family: Microsoft JhengHei;
  font-size: 18px;
}
.formContainer button {
  background-color: #CC9900;
  border-color: #775900;
  color: #FFF;
  width: 100px;
  margin-left: auto;
}
.formContainer button:disabled {
  background-color: #666666;
}
.formFieldFlex {
  display: flex;
  flex-direction: row;
  align-items: center;
}
.formFieldFlex input {
  font-family: Microsoft JhengHei;
  font-size: 18px;
}
.validationFlex {
  display: flex;
  flex-direction: row;
  align-items: center;
  height: 25px;
  margin-bottom: 5px;
}
.validationFlex div {
  width: 100px;
}
.validationFlex p {
  margin: 0px;
  margin-left: 20px;
  color: #FFF;
  font-style: italic;
}
.formFieldFlex input {
  width: 350px;
}

@media (min-width: 768px) and (max-width: 1024px) {
  .loginContainer {
    width: initial;
    background-color: #CC0000;
    padding-top: 20px;
  }
  .loginLayoutWidth {
    width: 90%;
    margin-left: auto;
    margin-right: auto;
  }
}

@media (max-width: 767.5px) {
  .formFieldFlex {
    flex-direction: column;
  }
  .formContainer input{
    margin-left: 0px;
    width: 80%;
  }
  .loginContainer {
    width: initial;
    height: 75%;
    background-color: #CC0000;
    padding-top: 20px;
  }
  .loginLayoutWidth {
    width: 90%;
    height: 90%;
    margin-left: auto;
    margin-right: auto;
    margin-top: 20px;
  }
  .validationFlex {
    flex-direction: column;
    align-items: center;
  }
  .validationFlex span {
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    width: 80%;
  }
  .validationFlex p {
    font-size: 14px;
    margin-left: 0px;
  }
  .formContainer label {
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    width: 80%;
    padding-right: 10px;
  }
  .formContainer label p {
    margin: 0px;
    margin-bottom: 5px;
  }
  .validationFlex div {
    display: none;
  }
  .formContainer button {
    margin-right: 5.5vw;
  }
}