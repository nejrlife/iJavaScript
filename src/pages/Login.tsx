import Header from "../components/Header";
import "./Login.less";
import { connect } from "react-redux";
import { useEffect, useState } from "react"
import { IS_AUTHENTICATED, LOG_IN } from "../sagas/constants";
import {
  useNavigate
} from "react-router-dom";

const Login = (props: any) => {
  const navigate = useNavigate();
  const [formData, setFormData] = useState({
    userId: '',
    password: '',
  });
  const [isUserIdValid, setUserIdValid] = useState(true);
  const [isPasswordValid, setPasswordValid] = useState(true);
  const [isButtonDisabled, setButtonDisabled] = useState(true);
  const [errorMessage, setErrorMessage] = useState('');

  const {
    dispatch  
  } = props;

  const handleInputChange = (event: any) => {
    const { name, value } = event.target;
    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const handleSubmit = (event: any) => {
    event.preventDefault(); // Prevent the default form submission behavior
    setErrorMessage('');
    if (formData.userId?.length === 0 || formData.password?.length === 0) {
      setErrorMessage('Please enter login credentials to continue');
      return;
    }
    // You can handle form submission logic here, e.g., send data to a server or perform client-side validation
    console.log('Form submitted with data:', event);
    dispatch({
      type: LOG_IN,
      payload: {
        formData
      }
    });
  };

  useEffect(() => {
    const token = sessionStorage.getItem('userToken');
    if (token && token?.length > 0) {
      dispatch({
        type: IS_AUTHENTICATED,
        payload: {
          userToken: token
        }
      });
    }
  }, [dispatch])

  useEffect(() => {
    if (props.isUserAuthenticated === true && props.userToken?.length > 0) {
      sessionStorage.setItem('userToken', props.userToken);
      navigate('/home/dashboard');
    }
  }, [props.isUserAuthenticated, props.userToken]);

  useEffect(() => {
    if (props?.loginError?.length > 0) {
      setErrorMessage(props.loginError);
    }
  }, [props.loginError]);

  useEffect(() => {
    const isUserValid = formData.userId?.length === 0 || formData.userId?.length >= 8 && formData.userId?.length <= 20;
    const isPwValid = formData.password?.length === 0 || formData.password?.length >= 8 && formData.password?.length <= 20;

    setUserIdValid(isUserValid);
    setPasswordValid(isPwValid);
    setButtonDisabled(!isUserValid || !isPwValid);
  }, [formData.userId, formData.password]);

  return (
    <div className='loginLayoutWidth'>
      <Header />
      <div className='loginContainer'>
        { errorMessage &&
          <div className='errorMessage'>
            <p><b>Error: </b>{ errorMessage }</p>
          </div> }
        <form onSubmit={handleSubmit}>
          <div className='formContainer'>
            <div className='formFieldFlex'>
              <label><p>User ID</p></label>
              <input
                type="text"
                placeholder="Enter User ID"
                name="userId"
                value={formData.userId}
                onChange={handleInputChange}
              />
            </div>
            <div className='validationFlex'>
              <div/>
              <span>{!isUserIdValid && <p>Please enter a valid username.</p>}</span>
            </div>
            <div className='formFieldFlex'>
              <label><p>Password</p></label>
              <input
                type="password"
                placeholder="Password"
                name="password"
                value={formData.password}
                onChange={handleInputChange}
              />
            </div>
            <div className='validationFlex'>
              <div/>
              <span>{!isPasswordValid && <p>Please enter a valid password.</p>}</span>
            </div>
            <button
              type="submit"
              className="basic-button"
              // disabled={isButtonDisabled}
            >Log In</button>
          </div>
        </form>
      </div>
    </div>
  )
}
const mapStateToProps = (state:any) => ({
  isUserAuthenticated: state.authenticateUser.isUserAuthenticated,
  userToken: state.authenticateUser.userToken,
  loginError: state.authenticateUser.logInError,
  customerId: state.authenticateUser.customerId
});
export default connect(mapStateToProps, null)(Login);