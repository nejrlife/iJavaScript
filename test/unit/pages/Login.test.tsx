import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { render, screen } from '@testing-library/react';
import Login from '../../../src/pages/Login/Login';
import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom'

const initAuthenticateUserDetailsState = {
  userToken: '',
  customerId: '',
  isUserAuthenticated: false,
  logInError: '',
  logInPending: false,
  isAuthenticatedError: '',
  isAuthenticatedPending: false
};

const INITIAL_STATE = {
  authenticateUser: { ...initAuthenticateUserDetailsState },
}

const mockStore = configureStore([]);

test('should render Login page and components', () => {
  const store = mockStore(INITIAL_STATE);
  const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </MemoryRouter>
  );
  const buttonElement = screen.getByText('Log In');
  const useridFieldElement = screen.getByPlaceholderText('Password');
  const pwfieldElement = screen.getByPlaceholderText('Enter User ID');
  
  expect(buttonElement).toBeInTheDocument();
  expect(useridFieldElement).toBeInTheDocument();
  expect(pwfieldElement).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test('Fix import React issue', () => {
  console.log(React);
});