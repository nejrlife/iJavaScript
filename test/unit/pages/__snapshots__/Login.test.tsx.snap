// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Login page 1`] = `
<div>
  <div
    class="loginLayoutWidth"
  >
    <div
      class="headerFlex red-box"
    >
      <p
        class="js_framework"
      >
        React JS
      </p>
      <p
        class="heading"
      >
        iJavaScript
      </p>
      <div
        class="logoutFlex"
      />
    </div>
    <div
      class="loginContainer"
    >
      <form>
        <div
          class="formContainer"
        >
          <div
            class="formFieldFlex"
          >
            <label>
              <p>
                User ID
              </p>
            </label>
            <input
              name="userId"
              placeholder="Enter User ID"
              type="text"
              value=""
            />
          </div>
          <div
            class="validationFlex"
          >
            <div />
            <span />
          </div>
          <div
            class="formFieldFlex"
          >
            <label>
              <p>
                Password
              </p>
            </label>
            <input
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="validationFlex"
          >
            <div />
            <span />
          </div>
          <button
            class="basic-button"
            type="submit"
          >
            Log In
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
